### ISSUE TITLE
`M0-1-10`:  Functions called from int32_t main() are not considered as "called"

### ISSUE THREAD

**Issue by rak3-sh** [github/codeql-coding-standards#646]
### Affected rules
 - M0-1-10

### Description
The `class MainFunction` defined in [EncapsulatingFunctions.qll](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/EncapsulatingFunctions.qll#L17) doesn't consider `int32_t main()` as a "main" function. Due to this, multiple false positives get raised in case the user has defined `main` with `int32_t` return type. 

### Example
Definitions of func1 and func2 are left out for reasons of brevity. 

```cpp
namespace mains {
    static int32_t var;

    // @brief namespace_func
    static void namespace_func(void) noexcept { // FP: Function is reported as not called, but it is called from "main" below.
        mains::var = -1;
        return;
    }
}  // namespace

// @brief main
// @return exit code
int32_t main(void) {
    int32_t ret {0};
    try {
        ret = func1(); // FP: Reported as not called in the definition location
        mains::var += ret;
        ret = func2(); // FP: Reported as not called in the definition location
        mains::var += ret;
    }
    catch(...) { 
        mains::namespace_func(); // namespace_func is called here.
    }
    return ret; 
}
```

**Comment by rak3-sh** [comment]
A possible fix strategy would be to consider Int32_t type as well for `MainFunction` [here](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/EncapsulatingFunctions.qll#L17).  It works locally. Happy to provide a PR that can be reviewed and integrated. Kindly let me know.

```
class MainFunction extends MainLikeFunction {
  MainFunction() {
    hasGlobalName("main") and
    (
      getType() instanceof IntType or
      getType() instanceof Int32_t
    )
  }
}
```

**Comment by lcartey** [comment]
Thanks for the report!

For a `main` function to be valid it needs to be using `int`, or an alias of `int`. So the above is only valid on platforms where `int32_t` is an alias for `int`. For this reason, I would suggest changing `MainFunction` to resolve typedefs, then perform the check against `IntType`, e.g.:
```codeql
class MainFunction extends MainLikeFunction {
  MainFunction() {
    hasGlobalName("main") and
    (
      getType().resolveTypeDefs() instanceof IntType
    )
  }
}
```
This would also have the benefit of capturing `main`s which use other aliases for `int`.

**Comment by rak3-sh** [comment]
I agree .. your proposal sounds better! Thank you for your comments!

**Comment by rak3-sh** [comment]
I was wondering how to modify the test for this rule. The test case uses the `int main` version. If I create another C++ file with `int32_t main`, will the test infrastructure treat it as a separate test? Should this be added in a separate directory?

**Comment by rak3-sh** [comment]
@lcartey : Would it be OK to have a test for this [here](https://github.com/github/codeql-coding-standards/tree/main/cpp/common/test/rules) considering that the fix is inside the library - [EncapsulatingFunctions.qll](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/EncapsulatingFunctions.qll#L17)?

**Comment by lcartey** [comment]
Library tests can be added to this directory:
https://github.com/github/codeql-coding-standards/tree/main/cpp/common/test/library/codingstandards/cpp

And, just to confirm, you will need a separate test directory for each type of `main` you want to test.

**Comment by rak3-sh** [comment]
@lcartey - thanks for your help as always! Kindly find the PR for the fix of this issue [here](https://github.com/github/codeql-coding-standards/pull/647).

**Comment by rak3-sh** [comment]
Issue fixed with PR: #647

**Comment by vergararamonrafol-lgtm** [comment]
[](https://www.facebook.com/c8ww111w4z)
